000-mm2-connector-offsets-support.md

```diff
@@ -0,0 +1,390 @@
+
+## Proposal
+
+Strimzi should include mechanisms to let users manage offsets when utilising the KafkaMirrorMaker2 resources.
+Many users of MirrorMaker make use of this resource, rather than the KafkaConnector resource, so it's important to support it here as well.
+Similar, to [proposal #76][proposal-76], this proposal suggests introducing a new annotation on the KafkaMirrorMaker2 resource to trigger each operation.
+The operations are:
+* List offsets.
+* Alter offsets.
+* Reset offsets.
+
+The list offsets operation uses a ConfigMap for the output.
+The alter offsets operation uses a ConfigMap for the input.
+
+In addition, a separate new annotation is proposed to allow the user to select which MirrorMaker connector to apply the action to.
+
+### Request/Response from the API endpoints
+
+The response body received when using the GET endpoint, and the request body used when calling the PATCH endpoint have identical structures to allow users to easily fetch and then alter offsets.
+The structure of the body is not fully defined for generic source connectors.
+This is because source connectors determine their own structure for offsets depending on the source system they are integrating with.
+However, for each of the MirrorMaker connectors (MirrorSourceConnector, MirrorCheckpointConnector, MirrorHeartbeatConnector) we know the expected structure.
+
+MirrorSourceConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "cluster": "east-kafka",
+        "partition": 0,
+        "topic": "mirrormaker2-cluster-configs"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorCheckpointConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "partition": 4,
+        "topic": "inventory",
+        "group": "my-group"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+MirrorHeartbeatConnector format:
+```
+{
+  "offsets": [
+    {
+      "partition": {
+        "sourceClusterAlias": "east-kafka",
+        "targetClusterAlias": "west-kafka"
+      },
+      "offset": {
+        "offset": 0
+      }
+    }
+  ]
+}
+```
+
+To alter offsets we do not need the user to supply the entirety of these JSON objects.
+That is because some of the fields are already known to the operator.
+Specifically, we can omit:
+* For MirrorSourceConnector:
+  * `cluster`, this is equal to `spec.mirrors[].sourceCluster` in the KafkaMirrorMaker2 resource
+  * `topic`, this is the configuration topic for the underlying Kafka Connect cluster, by default named `mirrormaker2-cluster-configs`
+* For MirrorCheckpointConnector no fields can be omitted
+* For MirrorHeartbeatConnector:
+  * `partition.sourceClusterAlias`, this is equal to `spec.mirrors[].sourceCluster` in the KafkaMirrorMaker2 resource
+  * `partition.targetClusterAlias`, this is equal to `spec.mirrors[].targetCluster` in the KafkaMirrorMaker2 resource
```

Comment:
* I wonder if it is wise to omit the information
    * The MM2 cluster might have different source clusters. So event through you seem to referece a specific source cluster in the annotation, it might be confusing once you have the Config Map with the data and don't know anymore what cluster it is from. So I would consider keeping it and validating it when altering the offsets?
    * We rely on the format as defined by MM2. Once you start modifying it, it might cause problems when Kafka changes the format. E.g. if new fields with the names you want to omit are added or the original fields are renamed. If you stick with the Kafka format as is, we will be in a much better position.
* If we decide to keep it, it might be nice to include examples of the actual structures you will use.

Reply:
Strimzi names the connectors as `<SOURCE_ALIAS>-><TARGET_ALIAS>.<CONNECTOR_NAME>` (e.g. `east-kafka->west-kafka.MirrorSourceConnector`) and since in the proposal the output is under an entry with that name I figured that would be sufficient to identify which specific connector the offsets are for, what do you think?

In terms of the format changing, we would still need a code change if we only validate it since the validation would start incorrectly failing, so is that a much better position?

I included examples where I explain how list and alter work for each connector type already, do you think I should also include it here in the proposal?

Reply:
> In terms of the format changing, we would still need a code change if we only validate it since the validation would start incorrectly failing, so is that a much better position?

My assumption is that the validation is only for a valid JSON as it is with the regular connectors. Although it is not completely clear from this proposal so it is probably worth clarifying it. That said, I think we should keep the validation behavior the same. That way we will not depend on Kafka so much for this.

The removal of the fields is tricky because if Kafka modifies it, how are you gonna decide what to do? Write another proposal to decide to remove some more fields? How will that affect the logic of adding it back to the JSON before passing it to Kafka if there are multiple versions on the Kafka side? That is what makes keeping it the same more future-proof for me.

> Strimzi names the connectors as <SOURCE_ALIAS>-><TARGET_ALIAS>.<CONNECTOR_NAME> (e.g. east-kafka->west-kafka.MirrorSourceConnector) and since in the proposal the output is under an entry with that name I figured that would be sufficient to identify which specific connector the offsets are for, what do you think?

TBH, I wondered if naming the fields in the Config Map based on connector name is the right choice. I did not comment about it because I decided that I'm not sure if it is a good or bad option. But relying on the field name only seems a risky business. That is also why I would keep it in the JSON.

Reply:
Ah I had assumed we were validating the keys we expected, that is what I intended when I stated
```
The data supplied by the user will be validated and augmented with information from the KafkaMirrorMaker2 resource before it is submitted to the Connect API.
The expected format will match the output from the list offsets action.

For example for a MirrorSourceConnector the ConfigMap `<CONNECTOR_NAME>.json` data field must contain an entry that matches:
...
```

but I can see that wording isn't as explicit as it could be.

The reason I suggested more complex validation behaviour is I figured that the KafkaMirrorMaker2 resource gives a more hand-holding experience for people running MirrorMaker. So it made sense to me that it also gives more help with offset management. I guess it would be a tricky thing for Kafka to evolve, because any changes to the structure you would expect to be backwards compatible, or the MM connectors would also need to be updated to handle the missing field.

Reply:
I personally would definitely aim for unification rather than for doing things differently.